interop-testing/src/test/java/io/grpc/testing/integration/NettyFlowControlTest.java

```diff
@@ -0,0 +1,315 @@
+
+  @AfterClass
+  public static void shutDownTests() {
+    executor.shutdown();
+  }
+
+  @After
+  public void endTest() throws IOException {
+    proxy.shutDown();
+    server.shutdown();
+  }
+
+  @Test
+  public void lowBandLowLatency() throws InterruptedException, ExecutionException {
+    resetProxy(LOW_BAND, LOW_LAT, TimeUnit.MILLISECONDS).get();
+    resetConnection(REGULAR_WINDOW, REGULAR_WINDOW);
+    doTest(LOW_BAND, LOW_LAT);
+  }
+
+  @Test
+  public void lowBandHighLatency() throws InterruptedException, ExecutionException {
+    resetProxy(LOW_BAND, HIGH_LAT, TimeUnit.MILLISECONDS).get();
+    resetConnection(REGULAR_WINDOW, REGULAR_WINDOW);
+    doTest(LOW_BAND, HIGH_LAT);
+  }
+
+  @Test
+  public void highBandLowLatency() throws InterruptedException, ExecutionException {
+    resetProxy(HIGH_BAND, LOW_LAT, TimeUnit.MILLISECONDS).get();
+    resetConnection(REGULAR_WINDOW, REGULAR_WINDOW);
+    doTest(HIGH_BAND, LOW_LAT);
+  }
+
+  @Test
+  public void highBandHighLatency() throws InterruptedException, ExecutionException {
+    resetProxy(HIGH_BAND, HIGH_LAT, TimeUnit.MILLISECONDS).get();
+    resetConnection(REGULAR_WINDOW, REGULAR_WINDOW);
+    doTest(HIGH_BAND, HIGH_LAT);
+  }
+
+  @Test
+  public void verySmallWindow() throws InterruptedException, ExecutionException {
+    resetProxy(MED_BAND, MED_LAT, TimeUnit.MILLISECONDS).get();
+    resetConnection(REGULAR_WINDOW, TINY_WINDOW);
+    doTest(MED_BAND, MED_LAT);
+  }
+
+  /**
+   * Main testing method. Streams 4 MB of data from a server and records the final window and
+   * average bandwidth usage
+   *
+   * @param bandwidth
+   * @param latency
+   */
+  private void doTest(int bandwidth, int latency) throws InterruptedException {
+
+    int streamSize = 4 * 1024 * 1024;
+
+    TestServiceGrpc.TestService stub = TestServiceGrpc.newStub(channel);
+    StreamingOutputCallRequest.Builder builder = StreamingOutputCallRequest.newBuilder();
+    builder.addResponseParameters(ResponseParameters.newBuilder().setSize(streamSize));
+    StreamingOutputCallRequest request = builder.build();
+
+    TestStreamObserver observer = new TestStreamObserver();
+    stub.streamingOutputCall(request, observer);
+    observer.waitFor();
+
+    // Don't actually need bw right now but keeping it around because it might be interesting
+    long bw = (streamSize / (observer.getElapsedTime() / TimeUnit.SECONDS.toNanos(1)));
+    long expectedWindow = (latency * (bandwidth / TimeUnit.SECONDS.toMillis(1))) * 2;
+    int lastWindow = windows.get(windows.size() - 1);
+
+    // deal with cases that either don't cause a window update or hit max window
+    expectedWindow = Math.min(MAX_WINDOW, (Math.max(expectedWindow, REGULAR_WINDOW)));
+
+    // Range looks large, but this allows for only one extra/missed window update
+    // (one extra update causes a 2x difference and one missed update causes a .5x difference)
+    assertTrue((lastWindow < 2 * expectedWindow) && (expectedWindow < 2 * lastWindow));
```

Comment:
You can use the assertThat call to break these into two separate asserts.

assertThat(lastWindow).IsLessThan(2 * expectedWindow)


That will give a better error message incase it fails.